File: sdks/go/handler.go
package emergent

import (
	"context"
	"time"
)

// EmergentHandler is a subscribe-and-publish primitive (transformation).
// Handlers subscribe to incoming messages, process them, and publish results.
type EmergentHandler struct {
	*baseClient
}

// ConnectHandler connects to the Emergent engine as a Handler.
func ConnectHandler(name string, opts *ConnectOptions) (*EmergentHandler, error) {
	client := newBaseClient(name, PrimitiveKindHandler, opts)
	if err := client.connect(opts); err != nil {
		return nil, err
	}
	return &EmergentHandler{baseClient: client}, nil
}

// Subscribe subscribes to the given message types and returns a MessageStream.
// The SDK automatically subscribes to system.shutdown and handles it internally.
func (h *EmergentHandler) Subscribe(ctx context.Context, messageTypes []string) (*MessageStream, error) {
	return h.subscribeInternal(ctx, messageTypes)
}

// Unsubscribe unsubscribes from the given message types.
func (h *EmergentHandler) Unsubscribe(ctx context.Context, messageTypes []string) error {
	return h.unsubscribeInternal(ctx, messageTypes)
}

// Publish publishes a message to the engine.
func (h *EmergentHandler) Publish(message *EmergentMessage) error {
	return h.publishInternal(message)
}

// PublishAck publishes a message and waits for broker acknowledgment.
// Unlike Publish, this provides backpressure by waiting for the engine's
// message broker to confirm it has processed and forwarded the message.
func (h *EmergentHandler) PublishAck(ctx context.Context, message *EmergentMessage) error {
	return h.publishInternalAck(ctx, message)
}

// PublishAll publishes all messages from a slice.
// Sends each message individually so subscribers begin consuming
// immediately. Stops on the first error.
// Returns the number of messages successfully published.
func (h *EmergentHandler) PublishAll(messages []*EmergentMessage) (int, error) {
	for i, msg := range messages {
		if err := h.PublishAck(context.Background(), msg); err != nil {
			return i, err
		}
	}
	return len(messages), nil
}

// PublishStream reads messages from a channel and publishes each one.
// Stops when the channel is closed, the context is cancelled, or
// a publish error occurs.
// Returns the number of messages successfully published.
func (h *EmergentHandler) PublishStream(ctx context.Context, ch <-chan *EmergentMessage) (int, error) {
	count := 0
	for {
		select {
		case <-ctx.Done():
			return count, ctx.Err()
		case msg, ok := <-ch:
			if !ok {
				return count, nil
			}
			if err := h.PublishAck(ctx, msg); err != nil {
				return count, err
			}
			count++
		}
	}
}

// StreamOffer offers items as a pull-based stream with consumer-driven backpressure.
// Publishes stream.ready, then serves items one at a time as the consumer
// sends stream.pull requests. Publishes stream.end when exhausted.
// Returns the number of items published.
func (h *EmergentHandler) StreamOffer(ctx context.Context, messageType string, items []map[string]any, pullStream *MessageStream, timeout time.Duration) (int, error) {
	streamID := generateCorrelationID("strm")

	// Announce stream
	ready, err := NewMessage("stream.ready")
	if err != nil {
		return 0, err
	}
	ready.WithPayload(map[string]any{"stream_id": streamID, "message_type": messageType, "count": len(items)})
	if err := h.Publish(ready); err != nil {
		return 0, err
	}

	published := 0
	idx := 0

	for {
		// Wait for pull request with timeout
		var msg *EmergentMessage
		select {
		case <-ctx.Done():
			return published, ctx.Err()
		case <-time.After(timeout):
			return published, &TimeoutError{Msg: "stream_offer timed out waiting for pull", Dur: timeout}
		case m, ok := <-pullStream.C():
			if !ok {
				return published, &ConnectionError{Msg: "pull stream closed during stream_offer"}
			}
			msg = m
		}

		// Check if it's a pull for this stream
		payload, _ := msg.Payload.(map[string]any)
		isPull := string(msg.MessageType) == "stream.pull" && payload != nil && payload["stream_id"] == streamID

		if isPull {
			if idx < len(items) {
				item, err := NewMessage(messageType)
				if err != nil {
					return published, err
				}
				item.WithPayload(items[idx]).WithMetadata(map[string]any{"stream_id": streamID})
				if err := h.Publish(item); err != nil {
					return published, err
				}
				published++
				idx++
			} else {
				end, err := NewMessage("stream.end")
				if err != nil {
					return published, err
				}
				end.WithPayload(map[string]any{"stream_id": streamID})
				if err := h.Publish(end); err != nil {
					return published, err
				}
				break
			}
		}
	}

	return published, nil
}

// StreamConsume consumes a pull-based stream, calling onItem for each item.
// Waits for stream.ready, then sends stream.pull requests automatically
// after each item is consumed. Stops on stream.end.
// Returns the number of items consumed.
func (h *EmergentHandler) StreamConsume(ctx context.Context, messageType string, sourceStream *MessageStream, timeout time.Duration, onItem func(*EmergentMessage)) (int, error) {
	// Wait for stream.ready
	var streamID string
	for {
		var msg *EmergentMessage
		select {
		case <-ctx.Done():
			return 0, ctx.Err()
		case <-time.After(timeout):
			return 0, &TimeoutError{Msg: "stream_consume timed out waiting for stream.ready", Dur: timeout}
		case m, ok := <-sourceStream.C():
			if !ok {
				return 0, &ConnectionError{Msg: "source stream closed before stream.ready"}
			}
			msg = m
		}

		if string(msg.MessageType) == "stream.ready" {
			payload, _ := msg.Payload.(map[string]any)
			if payload != nil && payload["message_type"] == messageType {
				if sid, ok := payload["stream_id"].(string); ok {
					streamID = sid
					break
				}
			}
		}
	}

	// Send initial pull
	pull, err := NewMessage("stream.pull")
	if err != nil {
		return 0, err
	}
	pull.WithPayload(map[string]any{"stream_id": streamID})
	if err := h.Publish(pull); err != nil {
		return 0, err
	}

	count := 0

	for {
		var msg *EmergentMessage
		select {
		case <-ctx.Done():
			return count, ctx.Err()
		case <-time.After(timeout):
			return count, &TimeoutError{Msg: "stream_consume timed out waiting for item", Dur: timeout}
		case m, ok := <-sourceStream.C():
			if !ok {
				return count, &ConnectionError{Msg: "source stream closed during stream_consume"}
			}
			msg = m
		}

		// Check for stream.end
		if string(msg.MessageType) == "stream.end" {
			payload, _ := msg.Payload.(map[string]any)
			if payload != nil && payload["stream_id"] == streamID {
				break
			}
		}

		// Check for matching data item
		meta, _ := msg.Metadata.(map[string]any)
		isItem := string(msg.MessageType) == messageType && meta != nil && meta["stream_id"] == streamID

		if isItem {
			onItem(msg)
			count++
			pull, err := NewMessage("stream.pull")
			if err != nil {
				return count, err
			}
			pull.WithPayload(map[string]any{"stream_id": streamID})
			if err := h.Publish(pull); err != nil {
				return count, err
			}
		}
	}

	return count, nil
}

// Discover queries the engine for available message types and primitives.
func (h *EmergentHandler) Discover(ctx context.Context) (*DiscoveryInfo, error) {
	return h.discoverInternal(ctx)
}

// SubscribedTypes returns the list of currently subscribed message types.
func (h *EmergentHandler) SubscribedTypes() []string {
	return h.subscribedTypesList()
}

// Name returns the handler's name.
func (h *EmergentHandler) Name() string {
	return h.name
}

// Close gracefully disconnects from the engine.
func (h *EmergentHandler) Close() error {
	return h.close()
}

File: sdks/ts/src/handler.ts
/**
 * EmergentHandler - Subscribe and publish client primitive.
 * @module
 */

import type {
  ConnectOptions,
  DiscoveryInfo,
  EmergentMessage,
} from "./types.ts";
import type { MessageBuilder } from "./message.ts";
import type { MessageStream } from "./stream.ts";
import { BaseClient } from "./client.ts";
import { createMessage } from "./message.ts";

/**
 * A Handler can both subscribe to and publish messages.
 *
 * Use this for processing components that transform, enrich, or route data.
 * Handlers are the workhorses of Emergent systems.
 *
 * @example
 * ```typescript
 * await using handler = await EmergentHandler.connect("order_processor");
 * await using stream = await handler.subscribe(["order.created"]);
 *
 * for await (const msg of stream) {
 *   const order = msg.payloadAs<Order>();
 *
 *   // Process and publish result with causation tracking
 *   await handler.publish(
 *     createMessage("order.processed")
 *       .causedBy(msg.id)
 *       .payload({ orderId: order.id, status: "ok" })
 *   );
 * }
 * ```
 */
export class EmergentHandler extends BaseClient
  implements Disposable, AsyncDisposable {
  private constructor(name: string, options?: ConnectOptions) {
    super(name, "Handler", options);
  }

  /**
   * Connect to the Emergent engine as a Handler.
   *
   * @param name - Unique name for this handler
   * @param options - Connection options
   *
   * @example
   * ```typescript
   * const handler = await EmergentHandler.connect("my_handler");
   * // ... use handler ...
   * handler.close();
   *
   * // Or with automatic cleanup:
   * await using handler = await EmergentHandler.connect("my_handler");
   * ```
   */
  static async connect(
    name: string,
    options?: ConnectOptions,
  ): Promise<EmergentHandler> {
    const handler = new EmergentHandler(name, options);
    await handler.connectInternal(options?.socketPath);
    return handler;
  }

  /**
   * Subscribe to message types and receive them via MessageStream.
   *
   * Supports both array and variadic arguments for convenience.
   *
   * @example
   * ```typescript
   * // Array style
   * const stream = await handler.subscribe(["order.created", "order.updated"]);
   *
   * // Variadic style
   * const stream = await handler.subscribe("order.created", "order.updated");
   *
   * for await (const msg of stream) {
   *   console.log(msg.messageType, msg.payload);
   * }
   * ```
   */
  async subscribe(
    typesOrFirst: string[] | string,
    ...rest: string[]
  ): Promise<MessageStream> {
    const types = Array.isArray(typesOrFirst)
      ? typesOrFirst
      : [typesOrFirst, ...rest];

    return await this.subscribeInternal(types);
  }

  /**
   * Unsubscribe from message types.
   *
   * @example
   * ```typescript
   * await handler.unsubscribe(["order.created"]);
   * ```
   */
  async unsubscribe(messageTypes: string[]): Promise<void> {
    await this.unsubscribeInternal(messageTypes);
  }

  /**
   * Publish a message.
   *
   * Supports multiple calling patterns for maximum ergonomics:
   *
   * @example
   * ```typescript
   * // 1. Shorthand: type + payload (most common)
   * await handler.publish("order.processed", { status: "ok" });
   *
   * // 2. With causation tracking (recommended for handlers)
   * await handler.publish(
   *   createMessage("order.processed")
   *     .causedBy(originalMsg.id)
   *     .payload({ status: "ok" })
   * );
   *
   * // 3. MessageBuilder (auto-calls .build())
   * await handler.publish(
   *   createMessage("order.processed")
   *     .causedBy(originalMsg.id)
   *     .payload({ status: "ok" })
   * );
   *
   * // 4. Complete EmergentMessage
   * await handler.publish(message);
   * ```
   */
  async publish(
    messageOrType: EmergentMessage | MessageBuilder | string,
    payload?: unknown,
  ): Promise<void> {
    let message: EmergentMessage;

    if (typeof messageOrType === "string") {
      // Shorthand: publish("type", { payload })
      message = createMessage(messageOrType).payload(payload).build();
    } else if (
      "build" in messageOrType && typeof messageOrType.build === "function"
    ) {
      // MessageBuilder: auto-call build()
      message = messageOrType.build();
    } else {
      // Already an EmergentMessage
      message = messageOrType as EmergentMessage;
    }

    await this.publishInternal(message);
  }

  /**
   * Publish a message with broker acknowledgment (backpressure).
   *
   * Unlike {@link publish}, this waits for the engine's message broker to
   * confirm it has processed and forwarded the message before returning.
   *
   * Used internally by {@link publishAll} and {@link publishStream}.
   */
  async publishAck(
    messageOrType: EmergentMessage | MessageBuilder | string,
    payload?: unknown,
  ): Promise<void> {
    let message: EmergentMessage;

    if (typeof messageOrType === "string") {
      message = createMessage(messageOrType).payload(payload).build();
    } else if (
      "build" in messageOrType && typeof messageOrType.build === "function"
    ) {
      message = messageOrType.build();
    } else {
      message = messageOrType as EmergentMessage;
    }

    await this.publishInternalAck(message);
  }

  /**
   * Publish all messages from an iterable.
   *
   * Sends each message individually with broker acknowledgment so
   * subscribers begin consuming immediately. Stops on the first error.
   *
   * @returns The number of messages successfully published.
   */
  async publishAll(
    messages: Iterable<EmergentMessage | MessageBuilder>,
  ): Promise<number> {
    let count = 0;
    for (const msg of messages) {
      await this.publishAck(msg);
      count++;
    }
    return count;
  }

  /**
   * Publish messages from an async iterable (stream).
   *
   * Consumes the async iterable, publishing each message individually with
   * broker acknowledgment so subscribers begin consuming immediately.
   * Stops on the first publish error or when the iterable ends.
   *
   * @returns The number of messages successfully published.
   */
  async publishStream(
    messages: AsyncIterable<EmergentMessage | MessageBuilder>,
  ): Promise<number> {
    let count = 0;
    for await (const msg of messages) {
      await this.publishAck(msg);
      count++;
    }
    return count;
  }

  /**
   * Offer items as a pull-based stream with consumer-driven backpressure.
   *
   * Publishes `stream.ready`, then serves items one at a time as the
   * consumer sends `stream.pull` requests. Publishes `stream.end`
   * when exhausted.
   *
   * @returns The number of items published
   */
  async streamOffer(
    messageType: string,
    items: Iterable<unknown> | AsyncIterable<unknown>,
    pullStream: MessageStream,
    timeout = 30000,
  ): Promise<number> {
    const streamId = `strm_${Date.now()}_${Math.random().toString(36).slice(2, 10)}`;

    // Announce stream
    await this.publish(
      createMessage("stream.ready").payload({
        stream_id: streamId,
        message_type: messageType,
      }),
    );

    // Create iterator
    const isAsync = Symbol.asyncIterator in Object(items);
    const iter = isAsync
      ? (items as AsyncIterable<unknown>)[Symbol.asyncIterator]()
      : (items as Iterable<unknown>)[Symbol.iterator]();

    let published = 0;

    while (true) {
      // Wait for pull request
      const msg = await Promise.race([
        pullStream.next(),
        new Promise<null>((_, reject) =>
          setTimeout(() => reject(new Error("stream_offer timed out waiting for pull")), timeout)
        ),
      ]);

      if (msg === null) break;

      const isPull =
        msg.messageType === "stream.pull" &&
        typeof msg.payload === "object" &&
        msg.payload !== null &&
        (msg.payload as Record<string, unknown>).stream_id === streamId;

      if (isPull) {
        const next = isAsync
          ? await (iter as AsyncIterator<unknown>).next()
          : (iter as Iterator<unknown>).next();

        if (!next.done) {
          await this.publish(
            createMessage(messageType)
              .payload(next.value)
              .metadata({ stream_id: streamId }),
          );
          published++;
        } else {
          await this.publish(
            createMessage("stream.end").payload({ stream_id: streamId }),
          );
          break;
        }
      }
    }

    return published;
  }

  /**
   * Consume a pull-based stream, yielding items one at a time.
   *
   * Waits for `stream.ready`, then sends `stream.pull` requests
   * automatically after each item is consumed. Iteration stops on
   * `stream.end`.
   */
  async *streamConsume(
    messageType: string,
    sourceStream: MessageStream,
    timeout = 30000,
  ): AsyncIterable<EmergentMessage> {
    // Wait for stream.ready
    let streamId: string | null = null;
    while (streamId === null) {
      const msg = await Promise.race([
        sourceStream.next(),
        new Promise<null>((_, reject) =>
          setTimeout(() => reject(new Error("stream_consume timed out waiting for stream.ready")), timeout)
        ),
      ]);

      if (msg === null) {
        throw new Error("Source stream closed before stream.ready");
      }

      if (
        msg.messageType === "stream.ready" &&
        typeof msg.payload === "object" &&
        msg.payload !== null &&
        (msg.payload as Record<string, unknown>).message_type === messageType
      ) {
        streamId = (msg.payload as Record<string, unknown>).stream_id as string;
      }
    }

    // Send initial pull
    await this.publish(
      createMessage("stream.pull").payload({ stream_id: streamId }),
    );

    // Yield items
    while (true) {
      const msg = await Promise.race([
        sourceStream.next(),
        new Promise<null>((_, reject) =>
          setTimeout(() => reject(new Error("stream_consume timed out waiting for item")), timeout)
        ),
      ]);

      if (msg === null) break;

      // Check for stream.end
      if (
        msg.messageType === "stream.end" &&
        typeof msg.payload === "object" &&
        msg.payload !== null &&
        (msg.payload as Record<string, unknown>).stream_id === streamId
      ) {
        break;
      }

      // Check for matching data item
      if (
        msg.messageType === messageType &&
        typeof msg.metadata === "object" &&
        msg.metadata !== null &&
        (msg.metadata as Record<string, unknown>).stream_id === streamId
      ) {
        yield msg;
        // Request next
        await this.publish(
          createMessage("stream.pull").payload({ stream_id: streamId }),
        );
      }
    }
  }

  /**
   * Discover available message types and primitives.
   *
   * @example
   * ```typescript
   * const info = await handler.discover();
   * console.log("Available types:", info.messageTypes);
   * ```
   */
  async discover(): Promise<DiscoveryInfo> {
    return await this.discoverInternal();
  }

  /**
   * Implement `Symbol.dispose` for `using` declaration support.
   */
  [Symbol.dispose](): void {
    this.close();
  }

  /**
   * Implement `Symbol.asyncDispose` for `await using` declaration support.
   */
  async [Symbol.asyncDispose](): Promise<void> {
    await this.disconnect();
  }
}
